File: webapps/fileservice/src/js/apis/filepermission.js
// Copyright (C) 2020 WuPeng <wup364@outlook.com>.
// Use of this source code is governed by an MIT-style.
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction,
// including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
// and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

"use strict";
import { $apitools } from "./apitools";
export const $filepms = {
	// 权限类型
	$TYPE: {
		// VISIBLECHILD 子节点可见导致的父节点可见, 父节点其实并没有权限
		VISIBLECHILD: { name: '子节点可见', value: 0 },
		// VISIBLE 可见
		VISIBLE: { name: '可见', value: 1 },
		// READ 只读
		READ: { name: '只读', value: 2 },
		// WRITE 可写
		WRITE: { name: '可写', value: 3 },
	},
	// listFPermissions
	listFPermissions() {
		return $apitools.apiGet("/filepms/v1/listfpermissions", {})
	},
	// GetUserPermissionSum
	GetUserPermissionSum(userid, paths) {
		return $apitools.apiPost("/filepms/v1/getuserpermissionsum", {
			userid: userid ? userid : '',
			paths: paths ? JSON.stringify(paths) : ''
		})
	},
	// listUserFPermissions
	listUserFPermissions(userid) {
		return $apitools.apiGet("/filepms/v1/listuserfpermissions", { userid: userid ? userid : '' })
	},
	// addFPermission
	addFPermission(userid, path, permission) {
		return $apitools.apiPost("/filepms/v1/addfpermission", {
			'userid': userid ? userid : '',
			'path': path ? path : '',
			'permission': permission ? permission : '',
		});
	},
	// delFPermission
	delFPermission(permissionid) {
		return $apitools.apiDelete("/filepms/v1/delfpermission", {
			'permissionid': permissionid ? permissionid : '',
		});
	},
	// updateFPermission
	updateFPermission(permissionid, permission) {
		return $apitools.apiPost("/filepms/v1/updatefpermission", {
			"permissionid": permissionid ? permissionid : '',
			"permission": permission ? permission : '',
		});
	},
	sync: {
		// delFPermission
		delFPermission(permissionid) {
			return $apitools.apiDeleteSync("/filepms/v1/delfpermission", {
				'permissionid': permissionid ? permissionid : '',
			});
		},
	}
};

// 权限类型计算
$filepms.$TYPE.__proto__ = {
	// 是否包含某权限
	sumInclude(sum, p) {
		sum = Number(sum); p = Number(p);
		if (undefined == sum || sum < 0 ||
			undefined == p || p < 0) {
			return false;
		}
		return 1 << p == (sum & (1 << p));
	},
	// 转换为描述文字
	sum2Name(sum) {
		let res = [];
		if (undefined != sum && sum > 0) {
			for (let key in $filepms.$TYPE) {
				if (key == $filepms.$TYPE.VISIBLECHILD.name ||
					!$filepms.$TYPE.hasOwnProperty(key)) {
					continue;
				}
				if ($filepms.$TYPE.sumInclude(sum, $filepms.$TYPE[key].value)) {
					res.push($filepms.$TYPE[key].name);
				}
			}
		}
		if (res.length == 0) {
			if ($filepms.$TYPE.sumInclude(sum, $filepms.$TYPE.VISIBLECHILD.value)) {
				res.push($filepms.$TYPE.VISIBLECHILD.name);
			}
		}
		return res;
	},
	// 计算权限结果值
	list2Sum(pms) {
		let res = -1;
		if (pms && pms.length > 0) {
			for (let i = 0; i < pms.length; i++) {
				res += 1 << Number(pms[i]);
			}
		}
		return res;
	},
};

File: server/biz/modules/filepermission/filepermission.go
// Copyright (C) 2020 WuPeng <wup364@outlook.com>.
// Use of this source code is governed by an MIT-style.
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction,
// including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
// and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

// 文件权限管理模块

package filepermission

import (
	"errors"
	"fileservice/biz/constants"
	"fileservice/biz/service"

	"github.com/wup364/pakku/ipakku"
	"github.com/wup364/pakku/utils/fileutil"
	"github.com/wup364/pakku/utils/logs"
)

// FilePermission 用户管理模块
type FilePermission struct {
	pmc  *PermissionCheck
	pms  *PermissionStory
	conf ipakku.AppConfig `@autowired:"AppConfig"`
}

// AsModule 模块加载器接口实现, 返回模块信息&配置
func (fpms *FilePermission) AsModule() ipakku.Opts {
	return ipakku.Opts{
		Name:        "FilePermission",
		Version:     1.0,
		Description: "文件权限模块",
		OnReady: func(mctx ipakku.Loader) {
			deftDataSource := "./.datas/" + mctx.GetParam(ipakku.PARAMKEY_APPNAME).ToString("app") + ".db?cache=shared"
			confDataSource := fpms.conf.GetConfig("filepermission.sotre.datasource").ToString(deftDataSource)
			if confDataSource == deftDataSource {
				fpms.mkSqliteDIR() // 创建sqlite文件存放目录
			}
			// PermissionStory
			fpms.pms = new(PermissionStory)
			if err := fpms.pms.Initial(constants.DBSetting{
				DriverName:     fpms.conf.GetConfig("filepermission.sotre.driver").ToString("sqlite3"),
				DataSourceName: confDataSource,
			}); nil != err {
				logs.Panicln(err)
			}
		},
		OnSetup: func() {
			// 执行建库、建表
			if err := fpms.pms.Install(); nil != err {
				logs.Panicln(err)
			}
		},
		OnInit: func() {
			// PermissionCheck
			fpms.pmc = NewPermissionCheck(fpms.pms)
			if err := fpms.pmc.InitPermission(); nil != err {
				logs.Panicln(err)
			}
		},
	}
}

// HashPermission 是否拥有某个权限
func (fpms *FilePermission) HashPermission(userID, path string, permission int64) bool {
	return fpms.pmc.HashPermission(userID, path, permission)
}

// GetUserPermissionSum 获取用户对路径拥有的权限数值
// -1 为无权限 > -1 为有权限
func (fpms *FilePermission) GetUserPermissionSum(userID, path string) int64 {
	return fpms.pmc.GetUserPermissionSum(userID, path)
}

// ListFPermissions 列出所有权限数据, 无分页
func (fpms *FilePermission) ListFPermissions() ([]service.PermissionInfo, error) {
	return fpms.pms.ListFPermissions()
}

// ListUserFPermissions 列出用户所有权限数据, 无分页
func (fpms *FilePermission) ListUserFPermissions(userID string) ([]service.PermissionInfo, error) {
	return fpms.pms.ListUserFPermissions(userID)
}

// QueryFPermission 根据权限ID查询详细信息
func (fpms *FilePermission) QueryFPermission(permissionID string) (*service.PermissionInfo, error) {
	return fpms.pms.QueryFPermission(permissionID)
}

// AddFPermission 添加权限
func (fpms *FilePermission) AddFPermission(fpm service.PermissionInfo) error {
	err := fpms.pms.AddFPermission(fpm)
	fpms.pmc.initPms2Memory()
	return err
}

// UpdateFPermission 修改权限
func (fpms *FilePermission) UpdateFPermission(fpm service.PermissionInfo) error {
	pms, err := fpms.pms.QueryFPermission(fpm.PermissionID)
	if nil != err {
		return err
	}
	if nil != pms {
		if pms.UserID == constants.AdminUserID {
			return errors.New("不能修改管理员权限")
		}
	}
	err = fpms.pms.UpdateFPermission(fpm)
	fpms.pmc.initPms2Memory()
	return err
}

// DelFPermission 根据permissionID删除权限
func (fpms *FilePermission) DelFPermission(permissionID string) error {
	pms, err := fpms.pms.QueryFPermission(permissionID)
	if nil != err {
		return err
	}
	if nil != pms {
		if pms.UserID == constants.AdminUserID {
			return errors.New("不能删除管理员权限")
		}
	}
	err = fpms.pms.DelFPermission(permissionID)
	fpms.pmc.initPms2Memory()
	return err
}

// mkSqliteDIR 创建sqlite文件存放目录
func (fpms *FilePermission) mkSqliteDIR() {
	if !fileutil.IsExist("./.datas") {
		if err := fileutil.MkdirAll("./.datas"); nil != err {
			logs.Panicln(err)
		}
	}
}
